src/coreComponents/constitutive/unitTests/testMultiFluid.cpp

```diff
@@ -251,12 +261,24 @@ void testNumericalDerivatives( MultiFluidBase & fluid,
       }
       compNew[jc] += dC;
 
-      // renormalize
-      real64 sum = 0.0;
-      for( localIndex ic = 0; ic < NC; ++ic )
-        sum += compNew[ic];
-      for( localIndex ic = 0; ic < NC; ++ic )
-        compNew[ic] /= sum;
+      // Note: in PVTPackage, derivatives are obtained with finite-difference approx
+      //       The component fraction is perturbed (just as above), and then all the component fractions are normalized (as below)
+      //       But, in the native DO model and in CO2Brine, derivatives are computed analytically. Say, in the 2-phase DO, that you want
+      //       to compute the derivatives of phase fraction wrt component fraction. Analytically we have
+      //          dPhaseFraction_dGlobalFraction[ic][jc] = (ic == jc)
+      //       which is correct but will fail the test below. Instead, the finite-difference approx a la PVTPackage yields
+      //          dPhaseFraction_dGlobalFraction[ic][jc] = (ic == jc) ? 1-composition[ic] : -composition[ic]
+      //       because of the normalization. The second option is, in my opinion, not correct, unless I missed some
+      //       cancellations that happen after when we differentiate wrt component densities. Hence the flag below.
+      if( usePVTPackage )
+      {
+        // renormalize
+        real64 sum = 0.0;
+        for( localIndex ic = 0; ic < NC; ++ic )
+          sum += compNew[ic];
+        for( localIndex ic = 0; ic < NC; ++ic )
+          compNew[ic] /= sum;
+      }
```

Comment:
Interesting issue. Though I don't fully understand the comment - you say the new (analytical) models have problems with this test, but you apply the re-normalization fix to the old (FD-based) models that used to work? Does this pass the test for both?

In any case, the problem seems to be applying a perturbation to a set of variables that aren't independent (constrained to unit sum). Perhaps the correct way to do this test (for both kinds of models) is to 1) apply the perturbation to one variable, 2) re-normalize, 3) compute "effective" deltas to each of the variables resulting from re-normalization, and 4) checking the numerical derivative w.r.t. all variables while taking into account the "effective" deltas (instead of `dC`, the original prescribed perturbation)?

Reply:
Yes sorry for the very unclear comment, and part of it is outdated. The issue comes from the normalization that is used in this test and in `PVTPackage` in the FD derivatives calculations (the GitHub diff is misleading here, I only added the flag `if( usePVTPackage )` to skip the normalization for the native models).

In `PVTPackage` the derivatives are obtained by 1) perturbing a given component fraction, and 2) re-normalizing so that the sum of the component fractions is equal to zero. With the normalization, the FD difference derivative **wrt component fraction** that is obtained is quite different from the analytical one. I tried to explain this for the phase fraction (equal to the component fraction in DO) in a two-phase model in the image below (not sure what GitHub will do with it):

![deriv_1](https://user-images.githubusercontent.com/20196406/111584078-9fb2ff00-877a-11eb-850e-ab4f5f92399c.png)
So as long as the normalization is applied in `PVTPackage`, it makes sense to have it for the models implemented there. My impression (not checked yet, will let you know) is that if we use an effective "delta", we are not going to match the `PVTPackage` derivative (Option A), but instead, we are going to get the analytical one (Option B).

In `develop`, the reason why the native DO model passes the test without the flag `if( usePVTPackage)` is that I enforced the "FD difference derivative with normalization" in my implementation here: https://github.com/GEOSX/GEOSX/blob/d4ade8e7be45494ea619e18980f39d1b75309b7d/src/coreComponents/constitutive/fluid/DeadOilFluid.hpp#L705
But I find this way of computing the derivatives not very intuitive, so now I would like to change it in this PR here: https://github.com/GEOSX/GEOSX/blob/2a038ad3548fc77b5aeec138e27738bfe4cc09a6/src/coreComponents/constitutive/fluid/DeadOilFluid.hpp#L707
and if the normalization is used to check derivatives, `testMultiFluid` says that my derivative is wrong (hence the flag)

Eventually, what we care about are the derivatives **wrt component density** that we put in the Jacobian, and both (normalization or no normalization) approaches are equivalent, even if the intermediate derivatives wrt component fractions are quite different. There is indeed a cancellation of terms taking place in the chain rule:
![deriv_2](https://user-images.githubusercontent.com/20196406/111587400-6761ef80-877f-11eb-8ab9-0221d2bccbad.png)

![deriv_3](https://user-images.githubusercontent.com/20196406/111587605-ae4fe500-877f-11eb-98df-1d9e08fc9b0e.png)

Reply:
Ok, I understand now, thanks for the detailed explanation! I agree that the new way of assigning `dPhaseFraction_dGlobalCompFraction` in DO is more intuitive (and mathematically correct, given that DO model itself does not apply input composition re-normalization, and leaves it to the user to adjust derivatives if they do so).

Reply:
For this normalization issue, in the latest commit I tried to improve my explanation in the inline comment just above the `if( usePVTPackage )` flag, and in this comment I reference this PR so people can view this discussion. For now, I left the flag that skips the normalization, but you are definitely right, this will have to be homogenized.